### PR TITLE
palette navigation

### PR DESCRIPTION
**Pull request by divya-urs**
The palette search results can be navigated using the arrow keys, and a palette element can be applied using the Enter key. These work if the keyboard focus is on the palette search box. There is a shortcut command to apply the currently selected palette element. This does not require keyboard focus to be on the search box. There is also a shortcut command to set focus to the palette search box.  On applying a palette element, focus shifts to the score.

### CODE REVIEW

File: mscore/masterpalette.h [hunk 1]
```diff
@@ -55,7 +55,7 @@ class MasterPalette : public QWidget, Ui::MasterPalette
       MasterPalette(QWidget* parent = 0);
       void selectItem(const QString& s);
       QString selectedItem();
-      };
```
Comment (MarcSabatella) [inline mscore/masterpalette.h]: FWIW this was correct as it was I think

File: mscore/palette.cpp [hunk 2]
```diff
@@ -91,11 +93,17 @@ Palette::Palette(QWidget* parent)
       setGrid(50, 60);
       _drawGrid     = false;
       _selectable   = false;
+      //if (!filterActive)
```
Comment (MarcSabatella) [inline mscore/palette.cpp]: if this code is definitely not needed, best to delete it at this point

File: mscore/palette.cpp [hunk 3]
```diff
@@ -136,12 +158,13 @@ bool Palette::filter(const QString& text)
                         }
                   if (c)
                         break;
-                  }
+                  }
```
Comment (MarcSabatella) [inline mscore/palette.cpp]: extra whitespace not needed?

File: mscore/palette.cpp [hunk 4]
```diff
@@ -136,12 +158,13 @@ bool Palette::filter(const QString& text)
                         }
                   if (c)
                         break;
-                  }
+                  }           
             if (t.isEmpty() || c)
                   dragCells.append(cell);
             bool contains = t.isEmpty() || c;
-            if (!contains)
+            if (!contains) {
```
Comment (MarcSabatella) [inline mscore/palette.cpp]: in general we don't use braces for single-line if's like this (also same thing a little bit above)

File: mscore/palette.cpp [hunk 5]
```diff
@@ -730,12 +765,306 @@ void Palette::leaveEvent(QEvent*)
...
+            currentIdx = i;
+            }
+      update();
+      }
+
+//---------------------------------------------------------
+//   applyPaletteElement
+//---------------------------------------------------------
+
+void Palette::applyPaletteElement()
```
Comment (MarcSabatella) [inline mscore/palette.cpp]: this is largely the same as mosueDoubleClickedEvent.  Is there a way to refactor the code to have a single function that does most of the work for both?

Comment (divya-urs) [inline mscore/palette.cpp]: I'm not sure. I think the definition of mouseDoubleClickedEvent can't be changed.

File: mscore/masterpalette.h (hunk 1 above)
Comment (divya-urs) [inline mscore/masterpalette.h]: I'll change this

File: mscore/palette.cpp (hunk 5 above)
Comment (MarcSabatella) [inline mscore/palette.cpp]: Anything can be changed :-).  The point would be to figure out how much code is truly shared by these two functions and then create a helper function that does the real work.  Then the duplicated code can be removed from both applyPaletteElement and mouseDoubleClickedEvent and replaced with a call to this new helper function.  Eg, mouseDoubleClickedEvent would have those few lines as the top processing the event itself to get "i".  After that pretty much everything else should be shareable, I would think.

**Comment by Jojo-Schmitz** [comment]
rebase needed...

File: mscore/events.cpp [hunk 6]
```diff
@@ -34,8 +34,11 @@ namespace Ms {
 
 bool ScoreView::event(QEvent* event)
       {
+
```
Comment (Jojo-Schmitz) [inline mscore/events.cpp]: empty line not needed

File: mscore/events.cpp [hunk 7]
```diff
@@ -483,7 +492,6 @@ struct CmdContext {
       CmdContext(Score* _s) : s(_s) { s->startCmd(); }
       ~CmdContext()                 { s->endCmd();   }
       };
-
```
Comment (Jojo-Schmitz) [inline mscore/events.cpp]: empty line needed

File: mscore/musescore.cpp [hunk 8]
```diff
@@ -1,4 +1,4 @@
-//=============================================================================
+﻿//=============================================================================
```
Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: Not sure what's the difference here?

File: mscore/palettebox.cpp [hunk 9]
```diff
@@ -364,7 +368,7 @@ void PaletteBox::write(XmlWriter& xml)
 
 QList<Palette*> PaletteBox::palettes()const
       {
-      QList<Palette*> pl;
+      QList<Palette*> pl;
```
Comment (Jojo-Schmitz) [inline mscore/palettebox.cpp]: extra whitespace not needed

File: mscore/palette.h [hunk 10]
```diff
@@ -134,13 +155,14 @@ class Palette : public QWidget {
...
       virtual void resizeEvent(QResizeEvent*) override;
+      void applyPaletteElement(PaletteCell* cell);
 
       virtual void dragEnterEvent(QDragEnterEvent*) override;
       virtual void dragMoveEvent(QDragMoveEvent*) override;
       virtual void dropEvent(QDropEvent*) override;
       virtual void contextMenuEvent(QContextMenuEvent*) override;
 
-      int idx(const QPoint&) const;
+
```
Comment (Jojo-Schmitz) [inline mscore/palette.h]: better delete the entire line

File: mscore/palette.cpp [hunk 11]
```diff
@@ -355,6 +382,7 @@ void Palette::mouseMoveEvent(QMouseEvent* ev)
                   currentIdx = -1;
             update();
             }
+
```
Comment (Jojo-Schmitz) [inline mscore/palette.cpp]: linefeed not needed

File: mscore/palettebox.h [hunk 12]
```diff
@@ -60,6 +65,14 @@ class PaletteBox : public QDockWidget {
...
       QList<Palette*> palettes() const;
       void updateWorkspaces();
+      QLineEdit* searchBox;
+      QLineEdit* getSearchBox();
+      bool noSelection();
+      void mousePressEvent(QMouseEvent* ev, Palette* p1);
+      void keyPressEvent(QKeyEvent *event);
+      void setKeyboardNavigation(bool val) { keyboardNavigation = val; }
+      bool getKeyboardNavigation() { return keyboardNavigation; }
+
```
Comment (Jojo-Schmitz) [inline mscore/palettebox.h]: linefeed too much

File: mscore/preferences.cpp [hunk 13]
```diff
@@ -1114,6 +1114,24 @@ void PreferenceDialog::updateSCListView()
...
+                        newSubItem->setData(0, Qt::UserRole, s->key());
+                        QString accessibleInfo = tr("Action: %1; Shortcut: %2")
+                           .arg(newSubItem->text(0)).arg(newSubItem->text(1).isEmpty()
+                              ? tr("No shortcut defined") : newSubItem->text(1));
+                        newSubItem->setData(0, Qt::AccessibleTextRole, accessibleInfo);
+                        newSubItem->setData(1, Qt::AccessibleTextRole, accessibleInfo);
+                        }
+                        continue;
+                  }*/
+
```
Comment (Jojo-Schmitz) [inline mscore/preferences.cpp]: dead code?

File: mscore/shortcutcapturedialog.cpp [hunk 14]
```diff
@@ -135,7 +156,6 @@ void ShortcutCaptureDialog::keyPress(QKeyEvent* e)
       // Check against conflicting shortcuts
       bool conflict = false;
       QString msgString;
-
```
Comment (Jojo-Schmitz) [inline mscore/shortcutcapturedialog.cpp]: I'd reather Keep this linefeed

File: mscore/musescore.cpp (hunk 8 above)
Comment (MarcSabatella) [inline mscore/musescore.cpp]: me neither.  something about the line feed maybe?

File: mscore/palettebox.h [hunk 15]
```diff
@@ -32,10 +34,13 @@ class PaletteBox : public QDockWidget {
...
       Palette* newPalette(const QString& name, int slot);
       QComboBox* workspaceList;
-      QLineEdit* searchBox;
+
+      //QMap<QString, Shortcut*> localShortcuts;
+
       const int paletteStretch = 1000;
       QAction* singlePaletteAction;
       QToolButton* addWorkspaceButton;
+      bool keyboardNavigation = false;
```
Comment (MarcSabatella) [inline mscore/palettebox.h]: This is basically the sense in which we now have "state" for the palette.  It's pretty lightweight and hopefully not a problem.  But since it was something that came up earlier (years ago) as a potential issue, it's worth pointing out here.

Basically, the code ties the keyboard navigation state to the search facility.  We enter this state after a search and then pressing up or down.  It gets cleared if you then apply an element (by hitting Enter).  Not sure if there are other places we should explicitly clear it.  I would think maybe if you clear the search.

Right now, the behavior after clearing the search is a little unclear to me.  If I search for "treble" and navigate through the matching clefs, then clear the search, then open the clefs palette normally, the clef I had navigated to still shows as selected, but doesn't *act* as if it is.  So hitting Enter won't apply it, and hitting Up or Down won't navigate.  So are we still in keyboardNavigation mode or not?  If not, I think we shouldn't still be highlighting the last-selected clef.

File: mscore/events.cpp [hunk 16]
```diff
@@ -36,6 +36,8 @@ bool ScoreView::event(QEvent* event)
       {
       if (event->type() == QEvent::KeyPress && editData.element) {
             QKeyEvent* ke = static_cast<QKeyEvent*>(event);
+            QKeyEvent keyPiano = QKeyEvent(QEvent::KeyPress, 80, Qt::ControlModifier, "Ctrl+Alt+P");
+            QString txt = ke->text();
```
Comment (lasconic) [inline mscore/events.cpp]: These two variables are not used right? Please delete if it's the case

File: mscore/events.cpp [hunk 17]
```diff
@@ -47,13 +49,19 @@ bool ScoreView::event(QEvent* event)
                         if (rv)
                               return true;
                         }
+
```
Comment (lasconic) [inline mscore/events.cpp]: Not needed.

File: mscore/scoreview.h [hunk 18]
```diff
@@ -174,6 +175,7 @@ class ScoreView : public QWidget, public MuseScoreView {
       virtual void mouseDoubleClickEvent(QMouseEvent*);
 
       virtual void keyPressEvent(QKeyEvent*) override;
+      void _keyPressEvent(QKeyEvent* ev);
```
Comment (lasconic) [inline mscore/scoreview.h]: What's the purpose of this ?

File: mscore/scoreview.h [hunk 19]
```diff
@@ -102,6 +102,7 @@ class ScoreView : public QWidget, public MuseScoreView {
 
       ViewState state;
       OmrView* _omrView;
+      QKeySequence key;
```
Comment (lasconic) [inline mscore/scoreview.h]: used?

File: mscore/scoreview.cpp [hunk 20]
```diff
@@ -1961,6 +1961,18 @@ void ScoreView::cmd(const QAction* a)
             else
                   cmdGotoElement(score()->lastElement());
             }
+      else if (cmd == "next-palette-element") {
+            MuseScore* ms = static_cast<MuseScore*>(this->parent());
```
Comment (lasconic) [inline mscore/scoreview.cpp]: mscore->getPaletteBox(); should be enough?
The rest could be put in a new method? PaletteBox::nextPaletteElement?

Comment (lasconic) [inline mscore/scoreview.cpp]: But is this really needed at all?

File: mscore/scoreview.h (hunk 18 above)
Comment (divya-urs) [inline mscore/scoreview.h]: I was trying things out in the beginning. I forgot to delete this, I'll delete it now.

File: mscore/scoreview.h (hunk 19 above)
Comment (divya-urs) [inline mscore/scoreview.h]: Not used. I've deleted it.

File: mscore/scoreview.cpp (hunk 20 above)
Comment (divya-urs) [inline mscore/scoreview.cpp]: This is not needed. I've deleted it.

File: mscore/scoreview.h (hunk 18 above)
Comment (divya-urs) [inline mscore/scoreview.h]: Sorry, I didn't write this function. It was there before. I haven't deleted it.

File: mscore/events.cpp [hunk 21]
```diff
@@ -492,15 +492,22 @@ void ScoreView::keyPressEvent(QKeyEvent* ev)
...
       editData.key       = ev->key();
       editData.modifiers = ev->modifiers();
       editData.s         = ev->text();
 
       if (MScore::debugMode)
             qDebug("keyPressEvent key 0x%02x(%c) mod 0x%04x <%s> nativeKey 0x%02x scancode %d",
                editData.key, editData.key, int(editData.modifiers), qPrintable(editData.s), ev->nativeVirtualKey(), ev->nativeScanCode());
+      }
 
+void ScoreView::_keyPressEvent(QKeyEvent* ev)
```
Comment (lasconic) [inline mscore/events.cpp]: I still don't get it. Why did you add _keyPressEvent??

File: mscore/musescore.cpp [hunk 22]
```diff
@@ -6267,6 +6278,11 @@ int main(int argc, char* av[])
       if (settings.value("mixerVisible", false).toBool())
             mscore->showMixer(true);
 
+      QAction* a = new QAction(getAction("palette-shortcut"));
+      QKeySequence keys = QKeySequence(Qt::Key_T);
```
Comment (lasconic) [inline mscore/musescore.cpp]: Let's not do that. Shortcuts are supposed to be set from the shortcuts.xml file. Ctrl + Shift + S could be used?

File: mscore/musescore.h [hunk 23]
```diff
@@ -717,6 +720,8 @@ class MuseScore : public QMainWindow, public MuseScoreCore {
       static Palette* newLinesPalette(PaletteType);
       static Palette* newFretboardDiagramPalette();
 
+      MasterPalette* getMasterPalette() {return masterPalette; }
```
Comment (lasconic) [inline mscore/musescore.h]: Do we need this?

File: mscore/palette.cpp [hunk 24]
```diff
@@ -819,6 +928,8 @@ static void paintPaletteElement(void* data, Element* e)
 
 void Palette::paintEvent(QPaintEvent* /*event*/)
       {
+      //if (filterActive)
```
Comment (lasconic) [inline mscore/palette.cpp]: If it's not useful, delete the lines.

**Comment by IsaacWeiss** [comment]
That is "Save As..."

File: mscore/musescore.cpp (hunk 22 above)
Comment (lasconic) [inline mscore/musescore.cpp]: Ok... it's "Save As...". Bad idea. Anyway, this should go in shortcuts.xml. We will find a shortcut.

File: mscore/events.cpp (hunk 21 above)
Comment (divya-urs) [inline mscore/events.cpp]: Sorry, I'll remove it.

File: mscore/musescore.h (hunk 23 above)
Comment (divya-urs) [inline mscore/musescore.h]: Not for this branch.

Comment (lasconic) [inline mscore/musescore.h]: then remove it too please.

File: mscore/palette.cpp [hunk 25]
```diff
@@ -730,12 +782,69 @@ void Palette::leaveEvent(QEvent*)
...
             update(r);
             }
       }
 
 //---------------------------------------------------------
+//   nextPaletteElement
+//---------------------------------------------------------
+void Palette::nextPaletteElement()
+      {
+      if (!filterActive)
```
Comment (lasconic) [inline mscore/palette.cpp]: Advanced workspace in French. Type "ba". You should see that the tremolo palette doesn't change but every element matches the filter but "filterActive" is false and so it's not possible to go though the palette.

Comment (MarcSabatella) [inline mscore/palette.cpp]: Good point.  Checking to see if the number of filtered cells is equal to the total number of cells is not a good enough way to tell if the filter is active or not.

File: mscore/events.cpp [hunk 26]
```diff
@@ -492,113 +492,16 @@ void ScoreView::keyPressEvent(QKeyEvent* ev)
       {
       if (state != ViewState::EDIT)
             return;
+
```
Comment (lasconic) [inline mscore/events.cpp]: no new line needed.

File: mscore/events.cpp [hunk 27]
```diff
@@ -492,113 +492,16 @@ void ScoreView::keyPressEvent(QKeyEvent* ev)
...
       CmdContext cmdContext(_score);
+      editData.view      = this;
       editData.key       = ev->key();
       editData.modifiers = ev->modifiers();
       editData.s         = ev->text();
 
       if (MScore::debugMode)
             qDebug("keyPressEvent key 0x%02x(%c) mod 0x%04x <%s> nativeKey 0x%02x scancode %d",
                editData.key, editData.key, int(editData.modifiers), qPrintable(editData.s), ev->nativeVirtualKey(), ev->nativeScanCode());
-
```
Comment (lasconic) [inline mscore/events.cpp]: We need all this code no ??

File: mscore/palette.cpp [hunk 28]
```diff
@@ -307,9 +332,15 @@ void Palette::mousePressEvent(QMouseEvent* ev)
                   }
             emit boxClicked(dragIdx);
             }
+      //currentIdx = idx(ev->pos());
```
Comment (lasconic) [inline mscore/palette.cpp]: If it's not needed, delete it

File: mscore/palettebox.cpp [hunk 29]
```diff
@@ -422,5 +427,145 @@ void PaletteBox::changeEvent(QEvent *event)
...
+      p1->setCurrentIdx(p1->idx(ev->pos()));
+      p1->update();
+      searchBox->setFocus();
+      }
+
+//---------------------------------------------------------
+//   keyPressEvent
+//---------------------------------------------------------
+
+void PaletteBox::keyPressEvent(QKeyEvent *event)
```
Comment (lasconic) [inline mscore/palettebox.cpp]: Be careful, keyPressEvent is a function inherited from QWidget http://doc.qt.io/qt-5/qwidget.html#keyPressEvent. Maybe you want to call this one differently??

File: mscore/palettebox.h (hunk 15 above)
Comment (MarcSabatella) [inline mscore/palettebox.h]: Currently, the behavior is more consistent - after applying an element, the highlighting is cleared, implying keyboard navigation really is disabled.  But maybe it would be better if it were consistent in the other direction - if keyboard navigation remained enabled.  So that returning to the search box put us right back where we were before, and could re-apply the same item again by hitting Enter again.

File: mscore/events.cpp (hunk 27 above)
Comment (divya-urs) [inline mscore/events.cpp]: Yes, I deleted it by mistake, sorry.

File: mscore/palettebox.cpp (hunk 29 above)
Comment (divya-urs) [inline mscore/palettebox.cpp]: Okay, I'll do that.

**Comment by lasconic** [comment]
Ok ! This is looking better and better !
Now I wonder, maybe we should change the shortcut to enter the symbol to Ctrl + Enter or something else and make it work also in the scoreview?
Use case:

I want to enter 4 quarter notes C D E F with a tremolo through the stem

* Press N
* Press C, press shortcut to go to the palette search box, type to show tremolo
* Press Cltr+Enter (for example)
* got a C with tremolo, 
* press D, Press Ctrl+Enter
* Press E, Press Ctrl+Enter
* Press F, Press Ctrl+Enter
Done !
Quite powerful!

Another question, while moving through the palette, the screen reader would work or not?

File: mscore/events.cpp [hunk 30]
```diff
@@ -517,7 +517,6 @@ void ScoreView::keyPressEvent(QKeyEvent* ev)
                   return;
                   }
             }
-
```
Comment (lasconic) [inline mscore/events.cpp]: add back this line please.
In fact, I just realized that you might wonder why we insist on not having new lines or blank lines being removed :) There is a good reason for it. Sometimes, we need to go back months in the git history to find a bug, and we use blame. Here is the blame view for this file https://github.com/musescore/MuseScore/blame/master/mscore/events.cpp
If you insert or delete lines, then, you will appear in the blame and it will be confusing for the person who will be trying to figure out a bug that has nothing to do with the commit message.

File: mscore/palette.h [hunk 31]
```diff
@@ -51,6 +55,8 @@ struct PaletteCell {
       double yoffset { 0.0   };      // in spatium units of "gscore"
       qreal mag      { 1.0   };
       bool readOnly  { false };
+      Palette* parent;
```
Comment (lasconic) [inline mscore/palette.h]: What's the purpose of "parent" here ? 
If possible, I would prefer to set the parent when creating the cell, and not having a getParent with a loop, and a method in the struct.

File: mscore/palette.h [hunk 32]
```diff
@@ -126,6 +145,8 @@ class Palette : public QWidget {
 
       bool _moreElements;
       bool _showContextMenu { true };
+      Preferences prefs;
+      QMap<QString, Shortcut*> localShortcuts;
```
Comment (lasconic) [inline mscore/palette.h]: Is it needed?

File: mscore/palette.h [hunk 33]
```diff
@@ -134,13 +155,13 @@ class Palette : public QWidget {
...
       virtual bool event(QEvent*) override;
       virtual void resizeEvent(QResizeEvent*) override;
+      void applyPaletteElement(PaletteCell* cell);
 
       virtual void dragEnterEvent(QDragEnterEvent*) override;
       virtual void dragMoveEvent(QDragMoveEvent*) override;
       virtual void dropEvent(QDropEvent*) override;
       virtual void contextMenuEvent(QContextMenuEvent*) override;
 
-      int idx(const QPoint&) const;
```
Comment (lasconic) [inline mscore/palette.h]: Why moved to the bottom?

File: mscore/palettebox.cpp [hunk 34]
```diff
@@ -54,6 +54,11 @@ PaletteBox::PaletteBox(QWidget* parent)
       setWidget(w);
 
       searchBox = new QLineEdit(this);
+      searchBox->setFocusPolicy(Qt::StrongFocus);
+      searchBox->installEventFilter(this);
+      //QLabel *searchLabel = new QLabel("&Search:", this);
```
Comment (lasconic) [inline mscore/palettebox.cpp]: Delete if not used, don't keep commented code.

File: mscore/palettebox.h [hunk 35]
```diff
@@ -60,6 +65,14 @@ class PaletteBox : public QDockWidget {
       void clear();
       QList<Palette*> palettes() const;
       void updateWorkspaces();
+      QLineEdit* searchBox;
+      QLineEdit* getSearchBox();
```
Comment (lasconic) [inline mscore/palettebox.h]: Do you really need getSearchbox ?

File: mscore/preferences.h [hunk 36]
```diff
@@ -190,6 +200,7 @@ struct Preferences {
       bool dirty;
 
       QList<PluginDescription> pluginList;
+      QMap<QString, PaletteCellDescription> paletteCells;
```
Comment (lasconic) [inline mscore/preferences.h]: Is this needed?

**Comment by MarcSabatella** [comment]
I really like the suggestion of making "apply palette element" available even with focus back in the score, great idea.  Right now, we do use Enter and Ctrl+Enter for apply line & page breaks,a nd of course just the other day I added code to make this more pervasive (doesn't require a barline to be selected).  But maybe we could have it both ways.  If there is a current selected palette item, we apply it, if not, we apply a break.  Something like that.

**Comment by lasconic** [comment]
I don't really like it. Blind users will have hard time to get if the palette is selected or not. So let's make it a shortcut and define it later?

File: mscore/preferences.h (hunk 36 above)
Comment (divya-urs) [inline mscore/preferences.h]: No, not for this branch

**Comment by MarcSabatella** [comment]
Good point.  So yes, a separate command to apply current palette element if any.  This still wouldn't prevent us from also allowing Enter to do the job while focus is in the palette, though, and I think that would feel natural.  So, either Enter or this new shortcut fro the palette to apply current item, then the new shortcut to reapply it fro the score.

File: mscore/preferences.h (hunk 36 above)
Comment (lasconic) [inline mscore/preferences.h]: Please delete then.

**Comment by lasconic** [comment]
Yes. Enter would work when searchbox has the focus. The other shortcut would work in all cases. I can live with that. @divya-urs let's focus on this, and clean this PR, let's merge it asap to gather some feedback. Very good job !

File: mscore/palettebox.h (hunk 35 above)
Comment (divya-urs) [inline mscore/palettebox.h]: No, it's not needed. I've deleted it.

**Comment by divya-urs** [comment]
Thanks @lasconic! I'll work on adding the new shortcut command now.

File: mscore/palette.h (hunk 31 above)
Comment (divya-urs) [inline mscore/palette.h]: It's not needed in this branch. I'll remove it.

File: mscore/palette.h (hunk 32 above)
Comment (divya-urs) [inline mscore/palette.h]: No, I'll delete it.

File: mscore/palette.h (hunk 33 above)
Comment (divya-urs) [inline mscore/palette.h]: It's needed in PaletteBox::mousePressEvent so that a palette element can be selected using the mouse.

File: mscore/musescore.h [hunk 37]
```diff
@@ -387,6 +388,7 @@ class MuseScore : public QMainWindow, public MuseScoreCore {
       QNetworkAccessManager* _networkManager { 0 };
       QAction* lastCmd                       { 0 };
       const Shortcut* lastShortcut           { 0 };
+      QKeySequence lastPaletteShortcut;
```
Comment (lasconic) [inline mscore/musescore.h]: Is that used ?

File: mscore/musescore.h [hunk 38]
```diff
@@ -34,6 +34,7 @@
 #include "libmscore/musescoreCore.h"
 #include "libmscore/score.h"
 #include "newwizard.h"
+#include "palettebox.h"
```
Comment (lasconic) [inline mscore/musescore.h]: is that used ?

File: mscore/palette.cpp [hunk 39]
```diff
@@ -1425,6 +1549,7 @@ QSize Palette::sizeHint() const
 void Palette::actionToggled(bool /*val*/)
       {
       selectedIdx = -1;
+      //int nn = size();
```
Comment (lasconic) [inline mscore/palette.cpp]: delete if not used please.

File: mscore/palette.cpp [hunk 40]
```diff
@@ -1512,6 +1637,28 @@ PaletteCellProperties::PaletteCellProperties(PaletteCell* p, QWidget* parent)
       }
 
 //---------------------------------------------------------
+//   PaletteShortcut
+//---------------------------------------------------------
+
+PaletteShortcut::PaletteShortcut(PaletteCell* p, QWidget* parent)
```
Comment (lasconic) [inline mscore/palette.cpp]: this is not used?

File: mscore/palette.cpp [hunk 41]
```diff
@@ -119,12 +122,34 @@ void Palette::resizeEvent(QResizeEvent* e)
 //---------------------------------------------------------
 
 bool Palette::filter(const QString& text)
-      {
+     {
```
Comment (IsaacWeiss) [inline mscore/palette.cpp]: Looks like a mistake.

Comment (divya-urs) [inline mscore/palette.cpp]: Yes, sorry, I'll remove that space

File: mscore/palette.cpp [hunk 42]
```diff
@@ -119,12 +122,34 @@ void Palette::resizeEvent(QResizeEvent* e)
...
+                  }
+            }
       bool res = true;
       dragCells.clear();
 
+      if (QString::compare(t, _name, Qt::CaseInsensitive) == 0) {
+            PaletteCell* c  = cells.first();
+            for (PaletteCell* cell : cells) {
+                  dragCells.append(cell);
+                  }
```
Comment (IsaacWeiss) [inline mscore/palette.cpp]: Per https://musescore.org/en/developers-handbook/musescore-coding-rules#Braces, no braces around a single line are preferred.

File: mscore/palette.cpp [hunk 43]
```diff
@@ -730,12 +781,71 @@ void Palette::leaveEvent(QEvent*)
...
+      PaletteBox* pb = mscore->getPaletteBox();
+      if (!pb->getKeyboardNavigation())
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i++;
+      if (i < size() && cellAt(i)) {
+            currentIdx = i;
+            }
```
Comment (IsaacWeiss) [inline mscore/palette.cpp]: Ditto.

File: mscore/palette.cpp [hunk 44]
```diff
@@ -730,12 +781,71 @@ void Palette::leaveEvent(QEvent*)
...
+      PaletteBox* pb = mscore->getPaletteBox();
+      if (!pb->getKeyboardNavigation())
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i--;
+      if (i >= 0 && cellAt(i)) {
+            currentIdx = i;
+            }
```
Comment (IsaacWeiss) [inline mscore/palette.cpp]: Ditto.

File: mscore/palette.cpp [hunk 45]
```diff
@@ -1022,7 +1132,12 @@ QPixmap Palette::pixmap(int paletteIdx) const
 
 bool Palette::event(QEvent* ev)
       {
-      if (columns() && (ev->type() == QEvent::ToolTip)) {
+      PaletteBox* pb = mscore->getPaletteBox();
+      if (filterActive && pb->getKeyboardNavigation() && (ev->type() == QEvent::MouseMove || ev->type() == QEvent::ToolTip
+          || ev->type() == QEvent::WindowDeactivate)) {
+            return true;
+            }
```
Comment (IsaacWeiss) [inline mscore/palette.cpp]: This one's good, though. ;-)

File: mscore/palette.cpp (hunk 41 above)
Comment (IsaacWeiss) [inline mscore/palette.cpp]: Or add it back... ;-)

Comment (divya-urs) [inline mscore/palette.cpp]: Yes, I've done that. I'll push the changes soon.

File: mscore/palette.h (hunk 33 above)
Comment (MarcSabatella) [inline mscore/palette.h]: So, changed from private to public, I guess?

File: mscore/musescore.cpp (hunk 8 above)
Comment (lasconic) [inline mscore/musescore.cpp]: Any chance this line can be fixed?

File: mscore/palette.cpp [hunk 46]
```diff
@@ -121,10 +122,32 @@ void Palette::resizeEvent(QResizeEvent* e)
...
+                  if (p->getCurrentIdx() != -1) {
+                        p->setCurrentIdx(-1);
+                        p->update();
+                        }
+                  }
+            }
       bool res = true;
       dragCells.clear();
 
+      if (QString::compare(t, _name, Qt::CaseInsensitive) == 0) {
```
Comment (lasconic) [inline mscore/palette.cpp]: Can you add a comment to explain what this block of code is about?

File: mscore/palette.h (hunk 33 above)
Comment (lasconic) [inline mscore/palette.h]: ok, thanks

File: mscore/musescore.cpp (hunk 8 above)
Comment (divya-urs) [inline mscore/musescore.cpp]: I'm not sure how to fix this. I tried deleting the line and pasting it again, but no change is detected.

Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: Just don't make that chunk part of the commit? Pretty easy in GitHub for Windows, just deselect that part of the changes, then commit, then ignore that remainder. Not sure how this works with git commandline though.

File: mscore/palette.cpp [hunk 47]
```diff
@@ -1022,7 +1130,12 @@ QPixmap Palette::pixmap(int paletteIdx) const
 
 bool Palette::event(QEvent* ev)
       {
-      if (columns() && (ev->type() == QEvent::ToolTip)) {
+      PaletteBox* pb = mscore->getPaletteBox();
```
Comment (lasconic) [inline mscore/palette.cpp]: Can you add a comment here to explain this long condition?

File: mscore/palettebox.h [hunk 48]
```diff
@@ -60,6 +61,13 @@ class PaletteBox : public QDockWidget {
       void clear();
       QList<Palette*> palettes() const;
       void updateWorkspaces();
+      QLineEdit* searchBox;
```
Comment (lasconic) [inline mscore/palettebox.h]: Keep this one private please. If you need to access it from outside, rename it to `_searchBox` and create an accessor `searchBox()`

File: mscore/palettebox.h (hunk 15 above)
Comment (lasconic) [inline mscore/palettebox.h]: It would be good to add a one line description of this boolean as a comment in the code.

File: mscore/musescore.cpp [hunk 49]
```diff
@@ -4599,6 +4599,26 @@ void MuseScore::cmd(QAction* a)
                QString("Command %1 not valid in current state").arg(cmdn));
             return;
             }
+      if (cmdn == "palette-search") {
+            PaletteBox* pb = getPaletteBox();
+            QLineEdit* sb = pb->searchBox;
+            sb->setFocus();
+            if (pb->noSelection())
```
Comment (lasconic) [inline mscore/musescore.cpp]: Are you aware that we will not pass here if the searchbox is used by just clicking on it and entering text? keyboardNavigation will not be changed then, is that a problem?

File: mscore/palette.cpp [hunk 50]
```diff
@@ -1045,6 +1158,15 @@ bool Palette::event(QEvent* ev)
                qApp->translate("Palette", cellAt(idx)->name.toUtf8()), this);
             return false;
             }
+      else if (ev->type() == QEvent::KeyPress) {
```
Comment (lasconic) [inline mscore/palette.cpp]: I still don't get why this is needed. QWidget::event() would the same (calling `keyPressEvent`) but it will return the result of `keyPressEvent`. Here you always return true and so stop the propagation of the event to the parent. Why is that needed again? Add a comment.

File: mscore/musescore.cpp (hunk 8 above)
Comment (divya-urs) [inline mscore/musescore.cpp]: Thanks, I think "git add -i" worked.

File: mscore/musescore.cpp (hunk 49 above)
Comment (divya-urs) [inline mscore/musescore.cpp]: Yes, this won't happen if the search box is clicked, but it's not a problem. If the search box is clicked, PaletteBox::eventFilter will be called, and when up/down/enter is pressed PaletteBox::navigation will be called, which sets keyBoardNavigation to true if there is no selection. So, it works even if the shortcut command for the search box is not used.

File: mscore/palette.cpp (hunk 50 above)
Comment (divya-urs) [inline mscore/palette.cpp]: I think it works even if this code is removed. I'll remove it.

File: mscore/musescore.cpp (hunk 8 above)
Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: Indeed, it is gone now.

File: mscore/palette.cpp (hunk 47 above)
Comment (divya-urs) [inline mscore/palette.cpp]: I've added a comment.
